CMake/unit_test.cmake

```diff
@@ -32,3 +32,23 @@ FUNCTION( ADD_UNIT_TEST TESTNAME PROCS THREADS TEST_BINARY )
     add_test_labels( ${TESTNAME} TEST_LABELS_TEMP )
     set_property(TEST ${TESTNAME} APPEND PROPERTY LABELS "unit")
 ENDFUNCTION()
+
+FUNCTION( ADD_NO_MPI_UNIT_TEST TESTNAME THREADS TEST_BINARY )
```

Comment:
If the code is build without MPI, ADD_UNIT_TEST handles the pure no MPI case.
If MPI is used, unit tests still needs to be built and run with MPI although only with 1 MPI rank. This is intentional because on HPC requires launcher (runjob/mpirun) to dispatch the execution on the compute node even for sequential code. 
So I don't see the need of this special treatment.

Reply:
If possible I don't want to run cmake again with different arguments to run a unit test. 

You don't need to link the message target, the mpi wrapper, the mpi libraries and all the problems with them and sanitizers to use jsrun to run a test. But you do need a different set of arguments.

But if you are not in an interactive job, which is the only way I run tests on summit. This then works fine.  Ctest is already controlling affinity and number of "jobs" It's also fine on workstations and the CI.

If we actually ran CI on summit via a submit script we would need to extend this ADD_NO_MPI_UNIT_TEST to include jsrun and a variable to pass the 
--smpiargs="-disable_gpu_hooks" which is needed to run non mpi test using CUDA with spectrum mpi but which in general I don't think we can pass to all our tests.
https://github.com/olcf/olcf-user-docs/pull/173/files

On most other HPC systems you could just use the mpirunner with 1 process and the test.

Reply:
> If possible I don't want to run cmake again with different arguments to run a unit test.

You can multiple build folders one with and another one without MPI. I saw this better than adding more function variants in our cmake. You only have to cmake once more.

> You don't need to link the message target, the mpi wrapper, the mpi libraries and all the problems with them and sanitizers to use jsrun to run a test. But you do need a different set of arguments.
 
you can run an interactive job and then ssh to the allocated compute node to run what you need for the non-MPI build.

> But if you are not in an interactive job, which is the only way I run tests on summit. This then works fine. Ctest is already controlling affinity and number of "jobs" It's also fine on workstations and the CI.

This is not fine. you may run things on the MOM nodes and got warnings by admins.

> If we actually ran CI on summit via a submit script we would need to extend this ADD_NO_MPI_UNIT_TEST to include jsrun and a variable to pass the
> --smpiargs="-disable_gpu_hooks" which is needed to run non mpi test using CUDA with spectrum mpi but which in general I don't think we can pass to all our tests.
> https://github.com/olcf/olcf-user-docs/pull/173/files

We will not run complicated CI on Summit. More friendly x86 is probably sufficient.

> On most other HPC systems you could just use the mpirunner with 1 process and the test.

That is true. But we do have machines like Theta, login/MOM and compute nodes are completely different architectures.

Sorry, I don't see a critical case that we must use this added function.

Reply:
I'll refine this today and PR, with summit and other systems with special mpi this is necessary.  Summit is the system that makes this necessary for me right now.

Your correct observation about the test code with regard to leaks and that I think all new code should pass sanitizers unambiguously in their unit tests now. This is needed. I of course know about multiple build folders but I don't see that as a good solution there is great devepoer hour cost to compilermany variants of the code inside of the edit, compile, debug, edit... loop.  I feel like (debug,  asan) * complex * precision is enough variants to compile at once. In my new code I am aiming at only requiring (debug, asan).

Reply:
I just realized that ADD_NO_MPI_UNIT_TEST is only a runner, it has nothing to do with linking the message target, the mpi wrapper. I think if you fix the mpirunner in ADD_NO_MPI_UNIT_TEST, it will probably just the same as ADD_UNIT_TEST

Reply:
That's what I just did :).  Looking in config though I'm not sure what will happen on theta.  That calls an aprunhelper script using the num procs argument. It might work, might not. Going to test on cades and summit.  Do you know of any other systems that have an issue like the cuda_hook library when a non mpi program is executed using jrsun, aprun, mpirun etc.?

Reply:
Are you saying you find ADD_NO_MPI_UNIT_TEST and ADD_UNIT_TEST becomes identical and we don't need to keep ADD_NO_MPI_UNIT_TEST?

Regarding cuda_hook library, summit is the only one I'm aware of but I'm not surprised HPC center needs it to detect user GPU usage.

Reply:
Accept on summit we don't need ADD_NO_MPI_UNIT_TEST.  On every system I've tried other than summit, the test executables that link `catch_main_no_mpi` instead of `catch_main` work fine with the same mpirunner setup as a 1 proc mpi executable.

Reply:
Yes. Link `catch_main_no_mpi` should serve your needs. CI indicates you last change here breaks no-MPI case. So Please remove ADD_NO_MPI_UNIT_TEST and use ADD_UNIT_TEST which does everything you need already IMO.

Reply:
Except on summit.

Reply:
Works on summit QMC_MPI=1 with flags as in build_olcf_summit_(Clang|xl).sh, works with openmpi on workstation QMC_MPI=1,0 with no special cmake flags.